Fix code to open Terraform resource documentation

dod38fr opened the pull request:
Currently the feature to open documentation is not always working. It relies on the output of `terraform providers` command. With AWS, this command works only with a valid token. This is a PITA when the token has a short life.

With this PR, the provider source is found by parsing current buffer and all terraform files for a `required_providers` declaration to find the provider source and build the documentation URL.

If the provider source is not found, this PR falls back on calling `terraform providers` like before.

One caveat: the `required_providers` declaration is searched with some regexps, so it's not bulletproof. Using a HCL parser or lsp would be better, but it's beyond my current skill set.

dod38fr committed:
add terraform--get-resource-provider-source-in-buffer function

dod38fr committed:
add terraform--get-resource-provider-source function

dod38fr committed:
can get provider source in otehr file (provider.tf)

dod38fr committed:
search provider source in all tf files

dod38fr committed:
open tf files when source if not found in current buffer

dod38fr committed:
fallback to "terrafrom providers" command

when provider source cannot be found in current buffer or in terraform
files

dod38fr committed:
update README.md

Fuco1 commented on test/test-command.el:
```diff
@@ -156,4 +156,49 @@ resource \"elasticstack_elasticsearch_security_user\" \"filebeat_writer\" {
+(ert-deftest command--terraform--get-resource-provider-source ()
+  (with-terraform-temp-buffer
+   "
+# blah blah
+terraform {
+  required_providers {
+    aws = {
+      source  = \"hashicorp/aws\"
+      version = \"~> 5\"
+    }
+    azurerm = {
+      source  = \"hashicorp/azurerm\"
+      version = \"~> 3.47\"
+    }
+  }
+}
```
Here we should probably put some different content so it really loads the data from the file and not from the buffer (as a fallback).

dod38fr replied:
Loading provider source from this file is done in test `command--terraform--get-resource-provider-source-provider-in-file` line 201.

Fuco1 commented on terraform-mode.el:
```diff
@@ -252,18 +252,54 @@
+(defun terraform--get-resource-provider-source (provider &optional dir)
+  "Return provider source for PROVIDER located in DIR."
+  (goto-char (point-min))
+  (let ((source) (file) (file_path) (tf_files))
```
Elisp never uses the underscore, traditional style is to use a dash to separate words.

Further, if you are only declaring variables, they can be written in a single paren, `(let (a b c d) ... )`.  They will all be set to nil.

But it's better to try to always declare and initialize at the same time rather than use `setq` later (kind of pretending we have immutability)

dod38fr replied:
ok for the first 2 points.

Originally, I did set `source` with `(let ...)`. But the other `*file*` variables can only be set if there are some files to read, which is not the case if the user has just started to write the terraform file and has not yet saved his buffer.

All in all, I found awkward to set `source` in `(let...)`, because it would be the only variable set this way, and it could be overridden later on.

dod38fr replied:
On 2nd thought, I've improved the code following your suggestion. Only `tf-files` is not setup in `let`

Fuco1 commented on terraform-mode.el:
```diff
@@ -252,18 +252,54 @@
         (match-string 1)))))
 
+(defun terraform--get-resource-provider-source (provider &optional dir)
+  "Return provider source for PROVIDER located in DIR."
```
It's not clear what the source is.  Is it buffers, files, file names?

dod38fr replied:
ok, I'll add these details in the doc.

Fuco1 commented on terraform-mode.el:
```diff
@@ -252,18 +252,54 @@
-                provider-ns
-                provider
+    (if (= (length provider-source) 0)
+        ;; fallback to old method with terraform providers command
+        (setq provider-source
+              (concat
+               (terraform--get-resource-provider-namespace provider)
+               "/" provider)))
+    (if (> (length provider-source) 0)
+        (format "https://registry.terraform.io/providers/%s/latest/docs/%s/%s"
+                provider-source
                 doc-dir
                 resource-name)
-      (user-error "Can not determine the provider namespace for %s" provider))))
+      (user-error "Can not determine the provider source for %s" provider))))
```
This code is a bit too complicated.  It's preferable not to re-assign variables in elisp if possible. 

When using only the `then` of the `if`, use `when` that will make the intent a bit clearer.

dod38fr replied:
Since there's 3 ways to set `provider-source` which are attempted, I don't know how to avoid re-assigning it. Any suggestion ?

Fuco1 reviewed (changes requested):
Thanks for the feature/fix.

This was originally very simple and now it's about 50 lines of code to implement.  It's getting a bit too complex :blush: 

There's some style suggestions and maybe some places the logic can be simplified, let's take it step by step.

dod38fr committed:
use kebab-case in variable names

"This is the way" in elisp world.

dod38fr committed:
remove extra () in let(...) statement

dod38fr committed:
update doc of terraform--get-resource-provider-source

dod38fr committed:
use 'when' instead of 'if' without 'else'

dod38fr committed:
rework and terraform--get-resource-provider-source

dod38fr commented:
@Fuco1 Any news ?

dod38fr committed:
avoid crashing on backup files

This function used to fail when a buffer is not saved. The message
was something like:

let*: Opening input file: Aucun fichier ou dossier de ce type, /home/domi/private/freelance/seira/repos/infra-terraform/modules/terragrunt/azure-config/.#cluster.tf

With this patch, terraform files name must be made with alphanum -
space and _ and finish with .tf

Fuco1 commented:
Thank you @dod38fr for following up on all my suggestions.  I am very grateful for your contribution and I'm sorry it took me so long to review and merge this.

dod38fr commented:
No worry. I'm glad it's been merged.

All the best
